lib/elixir/lib/map.ex

```diff
@@ -220,6 +220,120 @@ defmodule Map do
     |> :maps.from_list()
   end
 
+  @doc """
+  Ensures the first argument is a `map` with the given
+  keys and default values.
+
+  The second argument must be a list of `t:term`s, specifying
+  a given key, or tuples specifying a key and a default value.
+
+  If the map has only the given keys, it returns
+  `{:ok, map}` with default values applied. Otherwise it
+  returns `{:error, invalid_keys}` with invalid keys.
+
+  See also: `validate!/2`.
+
+  ## Examples
+
+      iex> Map.validate(%{}, [{"one", 1}, two: 2])
+      {:ok, %{"one" => 1, two: 2}}
+
+      iex> Map.validate(%{two: 3}, [one: 1, two: 2])
+      {:ok, %{one: 1, two: 3}}
+
+  If atoms are given, they are supported as keys but do not
+  provide a default value:
+
+      iex> Map.validate(%{}, [:one, two: 2])
+      {:ok, %{two: 2}}
+
+      iex> Map.validate(%{one: 1}, [:one, two: 2])
+      {:ok, %{one: 1, two: 2}}
+
+  Passing unknown keys returns an error:
+
+      iex> Map.validate(%{three: 3, four: 4}, [one: 1, two: 2])
+      {:error, [:three, :four]}
+  """
+  @doc since: "1.13.0"
+  @spec validate(map, [{term(), term()} | term()]) :: {:ok, map()} | {:error, [atom]}
+  def validate(map, values) when is_map(map) and is_list(values) do
+    # split values into the default half and the non-default half
+    {keys_with_defaults, non_default_keys} =
+      for val <- values, reduce: {[], []} do
+        {with_defaults, non_defaults} ->
+          case val do
+            {_, _} -> {[val | with_defaults], non_defaults}
+            _ -> {with_defaults, [val | non_defaults]}
```

Comment:
Should I only accept atoms and strings as the non-default keys? I've ended up relaxing this condition because maps are more versatile than keywords in this regard.